x/emissions/keeper/keeper.go

```diff
@@ -1004,38 +1018,50 @@ func (k *Keeper) AppendInference(
 		return types.ErrCantUpdateEmaMoreThanOncePerWindow
 	}
 
-	// append inference if not yet reached topN
-	if uint64(len(inferences.Inferences)) < moduleParams.MaxTopInferersToReward {
-		inferences.Inferences = append(inferences.Inferences, inference)
-		return k.InsertInferences(ctx, topic.Id, nonceBlockHeight, inferences)
-	}
-
-	// get score of current inference and check with lowest score
-	lowScore, lowScoreIndex, err := GetLowScoreFromAllInferences(ctx, k, topic.Id, inferences)
+	lowestEmaScore, found, err := k.GetLowestInfererScoreEma(ctx, topic.Id)
 	if err != nil {
-		return errorsmod.Wrap(err, "error getting low score from all inferences")
+		return errorsmod.Wrap(err, "error getting lowest inferer score ema")
+	} else if !found {
+		lowestEmaScore, err = GetLowestScoreFromAllInferers(ctx, k, topic.Id, workerAddresses)
+		if err != nil {
+			return errorsmod.Wrap(err, "error getting lowest score from all inferers")
+		}
+		err = k.SetLowestInfererScoreEma(ctx, topic.Id, lowestEmaScore)
+		if err != nil {
+			return errorsmod.Wrap(err, "error setting lowest inferer score ema")
+		}
 	}
-	// Update EMA score for the lowest score inferer
-	// This conditional decides who that lowest score inferer is
-	if previousEmaScore.Score.Gt(lowScore.Score) {
+
+	if previousEmaScore.Score.Gt(lowestEmaScore.Score) {	
 		// Update EMA score for the lowest score inferer, who is not the current inferer
 		err = k.CalcAndSaveInfererScoreEmaWithLastSavedTopicQuantile(
 			ctx,
 			topic,
 			nonceBlockHeight,
-			inferences.Inferences[lowScoreIndex].Inferer,
+			lowestEmaScore,
 		)
 		if err != nil {
 			return errorsmod.Wrap(err, "error calculating and saving inferer score ema with last saved topic quantile")
 		}
-		// Remove the lowest score inferer from the list
-		inferences.Inferences = append(inferences.Inferences[:lowScoreIndex], inferences.Inferences[lowScoreIndex+1:]...)
-		// Append the current inferer to the list
-		inferences.Inferences = append(inferences.Inferences, inference)
-		return k.InsertInferences(ctx, topic.Id, nonceBlockHeight, inferences)
+		// Remove inferer with lowest score
+		err = k.RemoveQualifiedInferer(ctx, topic.Id, lowestEmaScore.Address)
+		if err != nil {
+			return errorsmod.Wrap(err, "error removing qualified inferer")
+		}
+		// Add new qualified inferer
+		err = k.AddQualifiedInferer(ctx, topic.Id, inference.Inferer)
+		if err != nil {
+			return errorsmod.Wrap(err, "error adding qualified inferer")
+		}
+		// Calculate new lowest score with updated infererAddresses
+		err = UpdateLowestScoreFromInfererAddresses(ctx, k, topic.Id, workerAddresses, inference.Inferer, lowestEmaScore.Address)
+		if err != nil {
+			return errorsmod.Wrap(err, "error getting low score from all inferences")
+		}
```

Comment:
Just to confirm, the strategy is to keep in `inferences` all of the actors that have ever been active in this block, and then perform a selection of them at closing time of the nonce (`closeWorkerNonce`).
This strategy needs both actions to be in sync and so I think it should be documented.